### ISSUE TITLE
Pyright says type is the name of the variable rather than its actual type

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#1801]
Pyright says type is the name of the variable itself rather than its actual type when setting a class variable with generic type in `__init_subclass__`.  Example below:

```py
from typing import ClassVar, Generic, TypeVar

T = TypeVar("T", bound="A")


class A:
    pass


class B(Generic[T]):
    pass


class C(A, Generic[T]):
    b: ClassVar[B[T]]

    def __init_subclass__(cls, b: B[T]) -> None:
        super().__init_subclass__()
        cls.b = b


some_b = B["D"]()


class D(C["D"], b=some_b):
    # Argument of type "some_b" cannot be assigned to parameter "b" of type "B[D]" in function "__init_subclass__"
    #   Type "some_b" cannot be assigned to type "B[D]"
    pass
```

This *may* be related to https://github.com/microsoft/pyright/issues/1796, as changing the type in `some_b` and `D` to `A`, ie.

```py
some_b = B[A]()


class D(C[A], b=some_b):
    pass
```
will not produce the error. 

Even if this error doesn't exist as of 47a8954, I would expect pyright to report the actual type and not the name of the variable in the error message. So some other underlying issue may also be present here.

Tested on version `1.1.135`.

**Comment by erictraut** [comment]
What's happening here is that you've created a circular dependency that pyright is struggling to handle. It's similar to the other issue that you filed recently (https://github.com/microsoft/pyright/issues/1796). I was able to work around this other issue fairly easily, but this one is much more difficult, and I'm not inclined to fix it. I'm instead going to recommend that you modify your code to avoid creating this circular dependency.

The reason you're seeing `some_b` in the error message is that pyright is evaluating the type of `some_b` to determine whether it's a type alias or a simple variable (since the two are syntactically identical in Python). It does so by analyzing the type of the expression on the RHS: `B["D"]()`. To analyze this type, it needs to first analyze the forward-referenced symbol `D`. To do so, it analyzes the class declaration for `D`, including its base classes and other arguments. It first encounters the argument `C["A"]`, which causes it to recursively evaluate class `C`, `A`, and `T`. It then attempts to validate that the type argument `D` honors the bound type `A` for type variable `T`, but it cannot do this yet because `D`'s class hierarchy is not yet established. This is the first circular dependency we hit. My change for https://github.com/microsoft/pyright/issues/1796 works around this issue by noticing that `D` is not yet fully analyzed and skips the bound type check. Next, we come to the argument `b=some_b`. The type for `some_b` is not yet known because we're still in the middle of evaluating it. The default assumption in this case is that `some_b` is a type alias, which is why the error message is referring to the type by the name `some_b`.

Pyright previously included an explicit check and error message for cyclical dependencies of classes. I ended up removing it because it was flagging errors in some cases that otherwise wouldn't have resulted in any problem. In this particular case, that check would have been useful because it would have made it clear to you the cause of this error.

You may wonder why mypy is able to handle certain circular references that pyright stumbles on. The reason is that mypy is a multipass analyzer. It analyzes the entire file top to bottom multiple times until the type of every symbol is determined. Pyright, by contrast, is a "lazy" type analyzer. It evaluates the type of a symbol and evaluates (on demand) any types that it depends upon. That means it evaluates the type of each symbol only once. This is much more efficient, which is why pyright is so much faster than mypy. It's also really important for a type checker that's powering a language server because interactive editing features demand good performance. One of the only downsides of a lazy type analyzer is that there are a few cases where it cannot deal with cyclical dependencies that happen to work OK in a multi-pass analyzer. This is one of those rare cases.

You can work around this limitation in one of two ways:

1. Eliminate the use of the variable `some_b` and directly use `b=B["D"]()` within the `D`'s class declaration.
2. Add an explicit type annotation for `some_b` so it's clear upfront that it's not a type alias:
```python
some_b: B = B["D"]()
```

**Comment by kennipj** [comment]
Thanks for the insight, really appreciate it!

It's unfortunate this limitation exists, as the alternatives are not very appealing to me from an ergonomic standpoint. I will revisit this way of subclassing, maybe there is a better alternative.
